File: src/music_liked_sync/cache.py
import json
import sqlite3
import time
from dataclasses import asdict
from pathlib import Path
from typing import Sequence

from .models import Track
from .utils import normalize_key


class SyncCache:
    def __init__(self, path: Path) -> None:
        self.path = path
        self.path.parent.mkdir(parents=True, exist_ok=True)
        with self._connect() as conn:
            conn.execute("PRAGMA journal_mode=WAL")
            conn.execute(
                """
                CREATE TABLE IF NOT EXISTS matches (
                    direction TEXT NOT NULL,
                    source_key TEXT NOT NULL,
                    source_track_json TEXT NOT NULL,
                    target_track_json TEXT NOT NULL,
                    updated_at REAL NOT NULL,
                    PRIMARY KEY (direction, source_key)
                )
                """
            )
            conn.execute(
                """
                CREATE TABLE IF NOT EXISTS liked_tracks (
                    service TEXT NOT NULL,
                    source_id TEXT NOT NULL,
                    updated_at REAL NOT NULL,
                    PRIMARY KEY (service, source_id)
                )
                """
            )
            conn.execute(
                """
                CREATE TABLE IF NOT EXISTS library_cache (
                    service TEXT NOT NULL PRIMARY KEY,
                    tracks_json TEXT NOT NULL,
                    fetched_at REAL NOT NULL
                )
                """
            )
            conn.commit()

    def _connect(self) -> sqlite3.Connection:
        return sqlite3.connect(self.path)

    @staticmethod
    def _serialize_track(track: Track) -> str:
        return json.dumps(asdict(track), ensure_ascii=False)

    @staticmethod
    def _deserialize_track(payload: str) -> Track:
        data = json.loads(payload)
        return Track(
            title=str(data.get("title") or ""),
            artists=tuple(str(artist) for artist in (data.get("artists") or [])),
            source_id=str(data.get("source_id") or ""),
            duration_ms=data.get("duration_ms"),
            album=data.get("album"),
        )

    def store_match(self, direction: str, source: Track, target: Track) -> None:
        source_key = normalize_key(source.title, source.artists)
        now = time.time()
        with self._connect() as conn:
            conn.execute(
                """
                INSERT INTO matches(direction, source_key, source_track_json, target_track_json, updated_at)
                VALUES (?, ?, ?, ?, ?)
                ON CONFLICT(direction, source_key) DO UPDATE SET
                    source_track_json=excluded.source_track_json,
                    target_track_json=excluded.target_track_json,
                    updated_at=excluded.updated_at
                """,
                (
                    direction,
                    source_key,
                    self._serialize_track(source),
                    self._serialize_track(target),
                    now,
                ),
            )
            conn.commit()

    def get_match(self, direction: str, source: Track) -> Track | None:
        source_key = normalize_key(source.title, source.artists)
        with self._connect() as conn:
            row = conn.execute(
                "SELECT target_track_json FROM matches WHERE direction = ? AND source_key = ?",
                (direction, source_key),
            ).fetchone()
        if not row:
            return None
        try:
            return self._deserialize_track(row[0])
        except (TypeError, ValueError, json.JSONDecodeError):
            return None

    def mark_liked(self, service: str, source_id: str) -> None:
        self.mark_liked_many(service, [source_id])

    def mark_liked_many(self, service: str, source_ids: Sequence[str]) -> None:
        now = time.time()
        rows = [(service, source_id, now) for source_id in sorted(set(source_ids)) if source_id]
        if not rows:
            return
        with self._connect() as conn:
            conn.executemany(
                """
                INSERT INTO liked_tracks(service, source_id, updated_at)
                VALUES (?, ?, ?)
                ON CONFLICT(service, source_id) DO UPDATE SET
                    updated_at=excluded.updated_at
                """,
                rows,
            )
            conn.commit()

    def is_liked(self, service: str, source_id: str) -> bool:
        with self._connect() as conn:
            row = conn.execute(
                "SELECT 1 FROM liked_tracks WHERE service = ? AND source_id = ? LIMIT 1",
                (service, source_id),
            ).fetchone()
        return row is not None

    def store_library(self, service: str, tracks: Sequence[Track]) -> None:
        payload = json.dumps([asdict(track) for track in tracks], ensure_ascii=False)
        now = time.time()
        with self._connect() as conn:
            conn.execute(
                """
                INSERT INTO library_cache(service, tracks_json, fetched_at)
                VALUES (?, ?, ?)
                ON CONFLICT(service) DO UPDATE SET
                    tracks_json=excluded.tracks_json,
                    fetched_at=excluded.fetched_at
                """,
                (service, payload, now),
            )
            conn.commit()

    def get_library(self, service: str, max_age_seconds: float) -> list[Track] | None:
        if max_age_seconds <= 0:
            return None
        with self._connect() as conn:
            row = conn.execute(
                "SELECT tracks_json, fetched_at FROM library_cache WHERE service = ?",
                (service,),
            ).fetchone()
        if not row:
            return None
        tracks_json, fetched_at = row
        if (time.time() - float(fetched_at)) > max_age_seconds:
            return None
        try:
            raw_tracks = json.loads(tracks_json)
            if not isinstance(raw_tracks, list):
                return None
            return [self._deserialize_track(json.dumps(item, ensure_ascii=False)) for item in raw_tracks]
        except (TypeError, ValueError, json.JSONDecodeError):
            return None

File: internal/music_liked_sync/cache.go
package music_liked_sync

import (
	"database/sql"
	"encoding/json"
	"os"
	"path/filepath"
	"time"

	_ "modernc.org/sqlite"
)

type SyncCache struct {
	path string
	db   *sql.DB
}

func NewSyncCache(path string) (*SyncCache, error) {
	if err := os.MkdirAll(filepath.Dir(path), 0755); err != nil {
		return nil, err
	}

	db, err := sql.Open("sqlite", path)
	if err != nil {
		return nil, err
	}

	if _, err := db.Exec("PRAGMA journal_mode=WAL"); err != nil {
		return nil, err
	}

	queries := []string{
		`CREATE TABLE IF NOT EXISTS matches (
			direction TEXT NOT NULL,
			source_key TEXT NOT NULL,
			source_track_json TEXT NOT NULL,
			target_track_json TEXT NOT NULL,
			updated_at REAL NOT NULL,
			PRIMARY KEY (direction, source_key)
		)`,
		`CREATE TABLE IF NOT EXISTS liked_tracks (
			service TEXT NOT NULL,
			source_id TEXT NOT NULL,
			updated_at REAL NOT NULL,
			PRIMARY KEY (service, source_id)
		)`,
		`CREATE TABLE IF NOT EXISTS library_cache (
			service TEXT NOT NULL PRIMARY KEY,
			tracks_json TEXT NOT NULL,
			fetched_at REAL NOT NULL
		)`,
	}

	for _, q := range queries {
		if _, err := db.Exec(q); err != nil {
			return nil, err
		}
	}

	return &SyncCache{path: path, db: db}, nil
}

func (c *SyncCache) Close() error {
	return c.db.Close()
}

func (c *SyncCache) StoreMatch(direction string, source, target Track) error {
	sourceKey := NormalizeKey(source.Title, source.Artists)
	sourceJSON, _ := json.Marshal(source)
	targetJSON, _ := json.Marshal(target)
	now := float64(time.Now().UnixNano()) / 1e9

	_, err := c.db.Exec(`
		INSERT INTO matches(direction, source_key, source_track_json, target_track_json, updated_at)
		VALUES (?, ?, ?, ?, ?)
		ON CONFLICT(direction, source_key) DO UPDATE SET
			source_track_json=excluded.source_track_json,
			target_track_json=excluded.target_track_json,
			updated_at=excluded.updated_at
	`, direction, sourceKey, string(sourceJSON), string(targetJSON), now)
	return err
}

func (c *SyncCache) GetMatch(direction string, source Track) (*Track, error) {
	sourceKey := NormalizeKey(source.Title, source.Artists)
	var targetJSON string
	err := c.db.QueryRow("SELECT target_track_json FROM matches WHERE direction = ? AND source_key = ?", direction, sourceKey).Scan(&targetJSON)
	if err == sql.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}

	var track Track
	if err := json.Unmarshal([]byte(targetJSON), &track); err != nil {
		return nil, err
	}
	return &track, nil
}

func (c *SyncCache) MarkLiked(service, sourceID string) error {
	return c.MarkLikedMany(service, []string{sourceID})
}

func (c *SyncCache) MarkLikedMany(service string, sourceIDs []string) error {
	if len(sourceIDs) == 0 {
		return nil
	}
	now := float64(time.Now().UnixNano()) / 1e9

	tx, err := c.db.Begin()
	if err != nil {
		return err
	}
	defer tx.Rollback()

	stmt, err := tx.Prepare(`
		INSERT INTO liked_tracks(service, source_id, updated_at)
		VALUES (?, ?, ?)
		ON CONFLICT(service, source_id) DO UPDATE SET
			updated_at=excluded.updated_at
	`)
	if err != nil {
		return err
	}
	defer stmt.Close()

	for _, id := range sourceIDs {
		if id == "" {
			continue
		}
		if _, err := stmt.Exec(service, id, now); err != nil {
			return err
		}
	}

	return tx.Commit()
}

func (c *SyncCache) IsLiked(service, sourceID string) (bool, error) {
	var exists int
	err := c.db.QueryRow("SELECT 1 FROM liked_tracks WHERE service = ? AND source_id = ? LIMIT 1", service, sourceID).Scan(&exists)
	if err == sql.ErrNoRows {
		return false, nil
	}
	if err != nil {
		return false, err
	}
	return true, nil
}

func (c *SyncCache) StoreLibrary(service string, tracks []Track) error {
	payload, err := json.Marshal(tracks)
	if err != nil {
		return err
	}
	now := float64(time.Now().UnixNano()) / 1e9

	_, err = c.db.Exec(`
		INSERT INTO library_cache(service, tracks_json, fetched_at)
		VALUES (?, ?, ?)
		ON CONFLICT(service) DO UPDATE SET
			tracks_json=excluded.tracks_json,
			fetched_at=excluded.fetched_at
	`, service, string(payload), now)
	return err
}

func (c *SyncCache) GetLibrary(service string, maxAgeSeconds float64) ([]Track, error) {
	if maxAgeSeconds <= 0 {
		return nil, nil
	}

	var tracksJSON string
	var fetchedAt float64
	err := c.db.QueryRow("SELECT tracks_json, fetched_at FROM library_cache WHERE service = ?", service).Scan(&tracksJSON, &fetchedAt)
	if err == sql.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}

	if (float64(time.Now().UnixNano())/1e9 - fetchedAt) > maxAgeSeconds {
		return nil, nil
	}

	var tracks []Track
	if err := json.Unmarshal([]byte(tracksJSON), &tracks); err != nil {
		return nil, err
	}
	return tracks, nil
}
